pkg/webhooks/serviceaccount/serviceaccount.go

```diff
@@ -0,0 +1,223 @@
+		"deployer",
+	}
+	protectedNamespaces = regexp.MustCompile(managedNamespaces)
+	exceptionNamespaces = []string{
+		"openshift-logging",
+		"openshift-user-workload-monitoring",
+		"openshift-operators",
+	}
+)
+
+type serviceAccountWebhook struct {
+	s runtime.Scheme
+}
+
+// NewWebhook creates the new webhook
+func NewWebhook() *serviceAccountWebhook {
+	scheme := runtime.NewScheme()
+	err := admissionv1.AddToScheme(scheme)
+	if err != nil {
+		log.Error(err, "Fail adding admissionsv1 scheme to serviceAccountWebhook")
+		os.Exit(1)
+	}
+	err = corev1.AddToScheme(scheme)
+	if err != nil {
+		log.Error(err, "Fail adding corev1 scheme to serviceAccountWebhook")
+		os.Exit(1)
+	}
+
+	return &serviceAccountWebhook{
+		s: *scheme,
+	}
+}
+
+// Authorized implements Webhook interface
+func (s *serviceAccountWebhook) Authorized(request admissionctl.Request) admissionctl.Response {
+	return s.authorized(request)
+}
+
+func (s *serviceAccountWebhook) authorized(request admissionctl.Request) admissionctl.Response {
+	var ret admissionctl.Response
+
+	sa, err := s.renderServiceAccount(request)
+	if err != nil {
+		log.Error(err, "Couldn't render a service account from the incoming request")
+		return admissionctl.Errored(http.StatusBadRequest, err)
+	}
+
+	if isProtectedNamespace(request) && !isAllowedUserGroup(request) {
+		if request.Operation == admissionv1.Delete && !isAllowedServiceAccount(sa) {
+			log.Info(fmt.Sprintf("Deleting operation detected on proteced serviceaccount: %v", sa.Name))
+			ret = admissionctl.Denied(fmt.Sprintf("Deleting protected service account under namespace %v is not allowed", protectedNamespaces))
+			ret.UID = request.AdmissionRequest.UID
+			return ret
+		}
+	}
+
+	ret = admissionctl.Allowed("Request is allowed")
+	ret.UID = request.AdmissionRequest.UID
+	return ret
+}
+
+// renderServiceAccount render the serviceaccount object from the requests
+func (s *serviceAccountWebhook) renderServiceAccount(request admissionctl.Request) (*corev1.ServiceAccount, error) {
+	decoder, err := admissionctl.NewDecoder(&s.s)
+	if err != nil {
+		return nil, err
+	}
+	sa := &corev1.ServiceAccount{}
+
+	if len(request.OldObject.Raw) > 0 {
+		err = decoder.DecodeRaw(request.OldObject, sa)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return sa, nil
+}
+
+// isAllowedUserGroup checks if the user or group is allowed to perform the action
+func isAllowedUserGroup(request admissionctl.Request) bool {
+	if utils.SliceContains(request.UserInfo.Username, allowedUsers) {
+		return true
+	}
+
+	for _, group := range allowedGroups {
+		if utils.SliceContains(group, request.UserInfo.Groups) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// isProtectedNamespace checks if the request is going to operate on the serviceaccount in the
+// protected namespace list
+func isProtectedNamespace(request admissionctl.Request) bool {
+	ns := request.Namespace
+
+	if protectedNamespaces.Match([]byte(ns)) && !utils.SliceContains(ns, exceptionNamespaces) {
```

Comment:
Is the list of namespaces we want to protect by this webhook different from https://github.com/openshift/managed-cluster-validating-webhooks/blob/408009d2a597965df8ed020413aab8d9ec004793/pkg/config/namespaces.go#L11 ?

Reply:
thanks! updated to use the privilegedNamespace list in `config`